File: netlogo-gui/src/main/prim/threed/_turtlesat.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.threed

import org.nlogo.agent.{ Agent3D, AgentSet, AgentSetBuilder }
import org.nlogo.api.AgentException
import org.nlogo.core.AgentKind
import org.nlogo.nvm.{ Context, Reporter }

class _turtlesat extends Reporter {

  override def report(context: Context): AgentSet = {
    val dx = argEvalDoubleValue(context, 0)
    val dy = argEvalDoubleValue(context, 1)
    val dz = argEvalDoubleValue(context, 2)
    val patch =
      try context.agent.asInstanceOf[Agent3D].getPatchAtOffsets(dx, dy, dz)
      catch {
        case _: AgentException => return AgentSet.emptyTurtleSet
      }
    if (patch == null)
      AgentSet.emptyTurtleSet
    else {
      val agentSetBuilder = new AgentSetBuilder(AgentKind.Turtle, patch.turtleCount)
      val it = patch.turtlesHere.iterator
      while(it.hasNext) {
        val turtle = it.next()
        if (turtle != null)
          agentSetBuilder.add(turtle)
      }
      agentSetBuilder.build()
    }
  }
}

File: netlogo-gui/src/main/prim/etc/_turtlesat.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc;

import org.nlogo.agent.Agent;
import org.nlogo.agent.AgentSet;
import org.nlogo.agent.AgentSetBuilder;
import org.nlogo.core.AgentKindJ;
import org.nlogo.api.LogoException;
import org.nlogo.core.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.Reporter;

public final class _turtlesat
    extends Reporter {


  @Override
  public Object report(final Context context)
      throws LogoException {
    double dx = argEvalDoubleValue(context, 0);
    double dy = argEvalDoubleValue(context, 1);
    org.nlogo.agent.Patch patch = null;
    try {
      patch = context.agent.getPatchAtOffsets(dx, dy);
    } catch (org.nlogo.api.AgentException e) {
      return AgentSet.emptyTurtleSet();
    }
    if (patch == null) {
      return AgentSet.emptyTurtleSet();
    }
    AgentSetBuilder agentSetBuilder = new AgentSetBuilder(AgentKindJ.Turtle(), patch.turtleCount());
    for (org.nlogo.agent.Turtle turtle : patch.turtlesHere()) {
      if (turtle != null) {
        agentSetBuilder.add(turtle);
      }
    }
    return agentSetBuilder.build();
  }
}
